R/joins.R

```diff
@@ -55,25 +46,36 @@ join_dt <- function(op) {
 }
 
 #' @rdname join.tbl_dt
-inner_join.data.table <- join_dt({merge(x, y, by = by$x, allow.cartesian = TRUE)})
+inner_join.data.table <- join_dt({merge(x, y, by.x = by$x, by.y = by$y, all = FALSE, allow.cartesian = TRUE)})
 
 #' @rdname join.tbl_dt
-left_join.data.table  <- join_dt({merge(x, y, by = by$x, all.x = TRUE, allow.cartesian = TRUE)})
+left_join.data.table <- join_dt({merge(x, y, by.x = by$x, by.y = by$y, all.x = TRUE, allow.cartesian = TRUE)})
 
 #' @rdname join.tbl_dt
-right_join.data.table  <- join_dt(merge(x, y, by = by$x, all.y = TRUE, allow.cartesian = TRUE))
+right_join.data.table <- join_dt({merge(x, y, by.x = by$x, by.y = by$y, all.y = TRUE, allow.cartesian = TRUE)})
 
 #' @rdname join.tbl_dt
 semi_join.data.table  <- join_dt({
-  # http://stackoverflow.com/questions/18969420/perform-a-semi-join-with-data-table
-  w <- unique(x[y, which = TRUE, allow.cartesian = TRUE])
-  w <- w[!is.na(w)]
-  x[w]
+   y <- as.data.table(y)
```

Comment:
Actually, I was thinking about a simple function that computes `filter_y`.

Templating: We could also do something like

```
join_dt <- function(op, suffix = TRUE) {
  if (suffix) suffix_op <- ...
  else suffix_op <- NULL
  template <- substitute({
    ...
    suffix_op
    ...
    op
    ...
  })
  f <- ...
  if (!suffix) formals(f) <- formals(f)[names(formals(f)) != "suffix"]
  ...
}
```

Both solutions are feasible, but really look too complicated for the task at hand. In the end it seems simpler to admin some duplication and get rid of the templating mechanism. But this doesn't have to be in that PR.